### PR TITLE
Introduce safer layer of sugar for nsStyleUnion

### PR DESCRIPTION
**Pull request by Manishearth**
This routes (almost) all access to nsStyleUnion through a largely safe interface, CoordData.

It also introduces a corresponding Rust enum, CoordDataValue, which can be used when interacting with CoordData

LLVM should optimize the costs away in release mode. eddyb tested this a bit, and LLVM has no trouble threading matches together with inlining -- so all of the matches using enums here will have the same generated code as the old matches on the units.

Some unresolved questions:

Should I provide convenience methods like `set_coord`, `set_auto`, etc on CoordData? `.set_enum(CoordDataValue::Something)` should optimize down to the same thing, but the convenience methods look cleaner and won't load the optimizer as much.

Also, we're converting immutable references to mutable ones, which can be used to cause unsafety using some acrobatics. Perhaps a trait-based approach is better?
The issue is that in some places we only have a `&CoordData` (eg copy_from), but CoordData internally is `*mut`. It would be nice if CoordData could parametrize over its mutability, but Rust doesn't let us do that.

The alternate approach is to make CoordData a trait (also CoordDataMut). The trait requires you to implement `get_union()` and `get_unit()`, and gives you the rest of the functions for free. `nsStyleCoord` would directly implement both traits. `nsStyleSides` would have `data_at(idx)` and `data_at_mut(idx)` methods which return a struct `SidesData` containing a reference to the Sides and the index, which itself implements the `CoordData` and `CoordDataMut` trait (we need two traits here because there will have to be two `SidesData` structs).

I decided not to implement the traits approach first since it's pretty trivial to change this code to use traits, and the current design is more straightforward.

Thoughts?

r? @bholley

cc @emilio @heycam

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12521)

### CODE REVIEW

**Comment by Manishearth** [comment]
Oh, another issue is that we need to make all unit/union fields private and add unsafe accessors. This will need to be done through bindgen changes, which I'll need to look into.

For now, we're directly setting the mUnit/mValue fields only for z-index, and there's a justifying comment.

File: ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs [hunk 1]
```diff
@@ -4,51 +4,246 @@
...
+                Integer(i) => {
+                    *self.unit = eStyleUnit_Integer;
+                    *(*self.union).mInt.as_mut() = i;
+                }
+                Enumerated(i) => {
+                    *self.unit = eStyleUnit_Enumerated;
+                    *(*self.union).mInt.as_mut() = i as i32;
+                }
+                Calc(calc) => {
+                    *self.unit = eStyleUnit_Calc;
```
Comment (Manishearth) [inline ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs]: This line shouldn't be here, since we assert otherwise in set_calc_value.

**Comment by Manishearth** [comment]
I went ahead and made the traits changes. They're in a separate commit so that we can still pick the old way of doing it if we want to.

It might be possible to clean up some of the impls in that file with macros. Not sure if it's necessary though.

**Comment by Manishearth** [comment]
https://github.com/servo/rust-bindgen/pull/20 will make it possible to make this even more safe 😀

**Comment by bholley** [comment]
So much nicer!

r=me with those comments addressed, but please smoketest this a bit. :-)

@bors-servo delegate+

---

Reviewed 1 of 1 files at r1, 1 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[components/style/properties/gecko.mako.rs, line 371 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8WZYgmL-ZllRoNFZR:-KN8WZYhUHCVVDxII9f3:-700130789) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/components/style/properties/gecko.mako.rs#L371)):*
> ```Rust
>     pub fn set_${ident}(&mut self, v: longhands::${ident}::computed_value::T) {
>         v.0.width.to_gecko_style_coord(&mut self.gecko.${gecko_ffi_name}.data_at_mut(${x_index}));
>         v.0.width.to_gecko_style_coord(&mut self.gecko.${gecko_ffi_name}.data_at_mut(${y_index}));
> ```

Don't you mean height here?

---

*[components/style/properties/gecko.mako.rs, line 377 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8XCrQ5Ol9noHSUOSZ:-KN8XCrQ5Ol9noHSUOS_:1942902154) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/components/style/properties/gecko.mako.rs#L377)):*
> ```Rust
>         self.gecko.${gecko_ffi_name}.data_at_mut(${x_index})
>                   .copy_from(&other.gecko.${gecko_ffi_name}.data_at(${x_index}));
>         self.gecko.${gecko_ffi_name}.data_at_mut(${x_index})
> ```

And don't you want y_index here?

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 188 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN7PwAASOea_Jwvj85F-r2-188:-KN8PpEl4FPOJVqLS-Cd:-933196225) ([raw file](https://github.com/servo/servo/blob/6e9a68a0dbd3fb97321bb99d8219c4bf1c909937/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L188)):*
<details><summary><i>Previously, Manishearth (Manish Goregaokar) wrote…</i></summary>
> This line shouldn't be here, since we assert otherwise in set_calc_value.
</details>
I don't see where Gecko_SetStyleCoordCalcValue is implemented (maybe hasn't landed yet?), but I'd think we should only pass the union and let servo take care of setting the unit. But why do we need this function at all? Given that we're already reset()ing above, can't we just set the pointer and call the FFI function to addref?

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 53 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8NoQNqzkVhq7hAgFJ:-KN8NoQNqzkVhq7hAgFK:-2059875209) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L53)):*
> ```Rust
> impl<'a> CoordData for SidesData<'a> {
>     fn unit(&self) -> nsStyleUnit {
>         self.sides.mUnits[self.index]
> ```

Given that this isn't marked unsafe, I'm presuming that something in the rust type system guarantees that we can't have index > 3. Out of curiosity, how does that work?

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 122 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8O4i0F5akyz7aiSid:-KN8O4i1ODh1XgCghRi8:919063119) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L122)):*
> ```Rust
> 
> #[derive(Copy, Clone)]
> /// Enum representing the tagged union that is CoordData
> ```

The general commenting style we've been using so far uses periods in places like this (like Gecko does). Unless there's a rust convention to omit them, I'd rather be consistent. This appears in various places in this PR.

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 125 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8RLkTAKllBOJ2fsfG:-KN8RLkUUeTvrZg4r4uk:-1777896752) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L125)):*
> CoordDataValues

I'd think we should call this CoordDataValue?

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 177 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8P9pEP_vGzLVljaF6:-KN8P9pEP_vGzLVljaF7:2070323625) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L177)):*
> ```Rust
> 
>     #[inline(always)]
>     fn set_enum(&mut self, value: CoordDataValues) {
> ```

Given that 'enum' is one of the values of the enumerated type itself, I'd prefer to just call this something more generic like "set".

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 250 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8QE3KwniprmFkMiGd:-KN8QE3LfE2G7DecvvLF:745610963) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L250)):*
> ```Rust
>     #[inline]
>     unsafe fn as_calc_mut(&mut self) -> &mut nsStyleCoord_Calc {
>         transmute(*self.union().mPointer.as_mut() as *mut nsStyleCoord_Calc)
> ```

Can we debug_assert the unit here?

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 268 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8Qu41P0obY9jipgZW:-KN8Qu42HZTv9HlLwMBR:1199460954) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L268)):*
> ```Rust
> 
>     #[inline(always)]
>     fn as_enum(&self) -> CoordDataValues {
> ```

Same thing with the word 'enum' - maybe .get() or .as_value()? Something consistent with what we do for set.

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 296 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8RlKaW28mHgRYiR8U:-KN8RlKaW28mHgRYiR8V:1041976126) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L296)):*
> ```Rust
>     /// While this should usually be called with the unit checked,
>     /// it is not an intrinsically unsafe operation to call this function
>     /// with the wrong unit
> ```

Aside from the fact that we're interpreting potentially-garbage memory as a float? That seems bad enough that I would nix this comment and add debug_asserts to all these methods. We might as well just mark them unsafe, because the only callers are already in unsafe blocks.

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 319 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8T-MiNg3xaOrqF7OD:-KN8T-MiNg3xaOrqF7OE:-1763690659) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L319)):*
> as_calc

The distinction between as_calc and get_calc is weird. Maybe we should call these get_calc and get_calc_value respectively?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12521#-:-KN8ZLHypIbDOhUzeS-K:-513535849)*

**Comment by Manishearth** [comment]
Review status: 2 of 6 files reviewed at latest revision, 11 unresolved discussions.

---

*[components/style/properties/gecko.mako.rs, line 371 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8WZYgmL-ZllRoNFZR:-KNBENt2MNk2iGSoU5eo:292888000) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/components/style/properties/gecko.mako.rs#L371)):*
<details><summary><i>Previously, bholley (Bobby Holley) wrote…</i></summary>
> Don't you mean height here?
</details>
Yep. Copy-paste error :smile: 

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 188 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN7PwAASOea_Jwvj85F-r2-188:-KNBFSiCpfJ4rrwRtapt:-1655336080) ([raw file](https://github.com/servo/servo/blob/6e9a68a0dbd3fb97321bb99d8219c4bf1c909937/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L188)):*
<details><summary><i>Previously, bholley (Bobby Holley) wrote…</i></summary>
> I don't see where Gecko_SetStyleCoordCalcValue is implemented (maybe hasn't landed yet?), but I'd think we should only pass the union and let servo take care of setting the unit. But why do we need this function at all? Given that we're already reset()ing above, can't we just set the pointer and call the FFI function to addref?
</details>
I generally prefer calling constructors from C++ instead of emulating it in Rust. We can't just set the pointer because Rust doesn't auto-construct uninitialized values. Especially because in this case the refcounting fields need to be initialized too.

I also have to allocate here, which I want to avoid doing on the Rust side.

In general wherever an allocation or constructor is involved I'd like to wrap as much of the operation as possible in a single Gecko FFI call.

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 53 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8NoQNqzkVhq7hAgFJ:-KNBEVNVC22kLMLfe9ag:-1763590716) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L53)):*
<details><summary><i>Previously, bholley (Bobby Holley) wrote…</i></summary>
> Given that this isn't marked unsafe, I'm presuming that something in the rust type system guarantees that we can't have index > 3. Out of curiosity, how does that work?
</details>
It panics!

Since both the index and the length are known at compile time, this can get optimized away. 

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 122 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8O4i0F5akyz7aiSid:-KNBF0CgG3fluRSxpf_L:388639590) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L122)):*
<details><summary><i>Previously, bholley (Bobby Holley) wrote…</i></summary>
> The general commenting style we've been using so far uses periods in places like this (like Gecko does). Unless there's a rust convention to omit them, I'd rather be consistent. This appears in various places in this PR.
</details>
Rust convention is usually to have periods only when there are multiple sentences, but it's a loose one that's inconsistently followed

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 177 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8P9pEP_vGzLVljaF6:-KNBFLmM539g9Ph0KMmo:-20830276) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L177)):*
<details><summary><i>Previously, bholley (Bobby Holley) wrote…</i></summary>
> Given that 'enum' is one of the values of the enumerated type itself, I'd prefer to just call this something more generic like "set".
</details>
set_value, since set is already defined by gecko_values.rs

---

*[ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs, line 296 \[r3\]](https://reviewable.io:443/reviews/servo/servo/12521#-KN8RlKaW28mHgRYiR8U:-KNBHaEiaTuFc9nWNDli:-373221498) ([raw file](https://github.com/servo/servo/blob/afd05df5f370c20de6f8a8a9b4645e38b43cbb9e/ports/geckolib/gecko_bindings/sugar/ns_style_coord.rs#L296)):*
<details><summary><i>Previously, bholley (Bobby Holley) wrote…</i></summary>
> Aside from the fact that we're interpreting potentially-garbage memory as a float? That seems bad enough that I would nix this comment and add debug_asserts to all these methods. We might as well just mark them unsafe, because the only callers are already in unsafe blocks.
</details>
Generally Rust has a very specific definition of `unsafe`, and you're not supposed to use it for other invariants. I personally disagree with this philosophy though -- so I'll mark these unsafe :smile: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12521)*

**Comment by Manishearth** [comment]
@bors-servo r=bholley

I didn't address the `Gecko_SetStyleCoordCalcValue` one (reasons given in a comment above). If you still think it necessary, I'll do it in a separate PR coupled with a Gecko change to remove that function.

**Pull request merged**